### PR TITLE
Use SOAP interface for bag 2.0 extract 

### PR DESCRIPTION
**Pull request by RoelWKramer**
Use SOAP interface for bag 2.0 extract

- Soap interface handler
- Add support for databases to tests.
- Test importer from start
#24029

### CODE REVIEW

File: src/gobbagextract/database/connection.py [hunk 1]
```diff
@@ -25,7 +25,9 @@ def connect(force_migrate=False):
     global session, engine
 
     try:
-        print('Connectiong to DB')
+        db_str = f"{DATABASE_CONFIG['username']}@{DATABASE_CONFIG['host']}:" \
+                 f"{DATABASE_CONFIG['port']}/{DATABASE_CONFIG['database']}"
+        print('Connecting to DB ', db_str)
```
Comment (RoelWKramer) [inline src/gobbagextract/database/connection.py:30]: remove db_str as it exposes db credentials

File: src/gobbagextract/datastore/bag_extract.py [hunk 2]
```diff
@@ -55,19 +57,20 @@ class ElementFormatter:
...
-        gml_str = ET.tostring(elm).decode('utf-8')
+    @staticmethod
+    def _gml_to_wkt(elm: ElementTree.Element) -> str:
+        gml_str = ElementTree.tostring(elm).decode('utf-8')
         gml = ogr.CreateGeometryFromGML(gml_str)
         gml.FlattenTo2D()
         return gml.ExportToWkt()
 
-    def _flatten_nested_list(self, lst: list, key_prefix: str):
+    def _flatten_nested_list(self, lst: list, key_prefix: str) -> dict:
```
Comment (RoelWKramer) [inline src/gobbagextract/datastore/bag_extract.py:73]: dict misses types

File: src/gobbagextract/datastore/bag_extract.py [hunk 3]
```diff
@@ -109,46 +112,38 @@ def _flatten_dict(self, d: dict):
...
             'a/e': 'f',
             'g/h': [4, 5]
         }
-
-        :param d:
-        :return:
         """
-
-        def flatten(dct: dict):
+        def flatten(dct: dict) -> dict:
```
Comment (RoelWKramer) [inline src/gobbagextract/datastore/bag_extract.py:121]: param and return type: dict missing kv types

File: src/gobbagextract/mutations/afgifte.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import re
+from typing import NamedTuple, Optional
+import datetime as dt
+
+from gobbagextract.config import RX_GEMEENTE_DATE, RX_DATE
+from gobcore.exceptions import GOBException
+
+
+class Afgifte(NamedTuple):
```
Comment (RoelWKramer) [inline src/gobbagextract/mutations/afgifte.py:9]: Add docstrings

File: src/gobbagextract/mutations/bagextract.py [hunk 5]
```diff
@@ -1,140 +1,132 @@
...
+        if date.day < self.FULL_IMPORT_DAY:
+            date -= relativedelta(months=1)
+        return date.replace(day=self.FULL_IMPORT_DAY)
 
-    def restart_import(self, last_import: MutationImport) -> Tuple[ImportMode, str, datetime.date]:
+    def restart_import(self, last_import: MutationImport) -> tuple[ImportMode, Afgifte, dt.date]:
         mode = last_import.mode
-        date, gemeente = self._date_gemeente_from_filename(last_import.filename)
+        afgifte = Afgifte(Bestandsnaam=last_import.filename)
+        date, gemeente = afgifte.get_date(), afgifte.get_gemeente()
```
Comment (RoelWKramer) [inline src/gobbagextract/mutations/bagextract.py:47]: 2 lines

File: src/gobbagextract/mutations/bagextract.py [hunk 6]
```diff
@@ -1,140 +1,132 @@
...
 
         if mode == ImportMode.FULL.value:
-            ret = self.start_full(date, gemeente)
+            ret = self.get_full(date, gemeente)
```
Comment (RoelWKramer) [inline src/gobbagextract/mutations/bagextract.py:50]: return statements here, remove else:

return self.get_full(date, gemeente), date

self.get_daily_mutations(date), date

File: src/gobbagextract/mutations/bagextract.py [hunk 7]
```diff
@@ -1,140 +1,132 @@
...
-            ret = self.start_mutations(date)
+            ret = self.get_daily_mutations(date)
 
-        return ret + (date, )
+        return ret + (date,)
 
-    def start_next(self, last_import: MutationImport, gemeente: str) -> Tuple[ImportMode, str, datetime.date]:
-        date, _ = self._date_gemeente_from_filename(last_import.filename)
-        next_date = date + datetime.timedelta(days=1)
+    def start_next(self, last_import: MutationImport, gemeente: str) -> tuple[ImportMode, Afgifte, dt.date]:
```
Comment (RoelWKramer) [inline src/gobbagextract/mutations/bagextract.py:56]: return statement in condition (same as above)

File: src/gobbagextract/mutations/bagextract.py [hunk 8]
```diff
@@ -1,140 +1,132 @@
...
-        if fname not in self._get_available_full_downloads(gemeente):
-            # TODO: This implies when file is not yet availble on the 15th we
-            # are failing this workflow!
-            raise NothingToDo.file_not_available(fname)
+        for afgifte in self.SoapHandler(**kwargs).iter_afgiftes():
+            if afgifte.get_gemeente() == gemeente and afgifte.get_date() == date:
+                return ImportMode.FULL, afgifte
 
-        return ImportMode.FULL, fname
+        # TODO: This implies when file is not yet availble on the 15th we are failing this workflow!
```
Comment (RoelWKramer) [inline src/gobbagextract/mutations/bagextract.py:80]: What should be fixed then? Log error instead of raising exception? 
This method is called in multiple places, so these places should handle this exception appropriately. In that case the todo should be moved to those places. Or maybe this is correct already.

File: src/gobbagextract/mutations/productstore.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+import requests
+
+from gobbagextract.config import KADASTER_PRODUCTSTORE_AFGIFTE_URL, KADASTER_PRODUCTSTORE_CERT, \
+    KADASTER_PRODUCTSTORE_KEY, \
+    KADASTER_PRODUCTSTORE_DOWNLOAD_URL
+from gobbagextract.mutations.afgifte import Afgifte
+
+
+class ProductStore:
```
Comment (RoelWKramer) [inline src/gobbagextract/mutations/productstore.py:13]: Add docstrings

File: src/gobbagextract/prepare/prepare_client.py [hunk 10]
```diff
@@ -10,6 +10,21 @@
 from gobbagextract.datastore.bag_extract import BagExtractDatastore
 
 
+def connect(func):
```
Comment (RoelWKramer) [inline src/gobbagextract/prepare/prepare_client.py:13]: Is this decorator used multiple times?
+ missing docstring, particulary important here.

File: src/gobbagextract/prepare/prepare_client.py [hunk 11]
```diff
@@ -18,18 +33,21 @@ class PrepareClient:
             {'name': 'object', 'type': 'JSON'},
     ]
 
-    def __init__(self, msg: dict, dataset, mode: ImportMode, last_date: datetime):
-        self.dataset = dataset
+    def __init__(self, msg: dict, dataset, mode: ImportMode, last_date: dt.date):
```
Comment (RoelWKramer) [inline src/gobbagextract/prepare/prepare_client.py:36]: dataset missing type

File: src/gobbagextract/selector/_selector.py [hunk 12]
```diff
@@ -27,47 +27,40 @@ def __init__(self, src_datastore: Datastore, dst_datastore: Datastore, config: d
...
+            rows = self._read_rows(self.query)
             chunk = itertools.islice(rows, self.WRITE_BATCH_SIZE)
-            columns = self.destination_table['columns']
             values = self._values_list(chunk, columns)
-            column_names = [c['name'] for c in columns]
-            self._write_rows(self.destination_table['name'], values, column_names)
+            self._write_rows(table, values, column_names)
 
             total_cnt += len(values)
+            # logger.info(f"Written {total_cnt:,} rows to destination table {table}")
```
Comment (RoelWKramer) [inline src/gobbagextract/selector/_selector.py:43]: remove this line

File: src/gobbagextract/selector/_selector.py [hunk 13]
```diff
@@ -27,47 +27,40 @@ def __init__(self, src_datastore: Datastore, dst_datastore: Datastore, config: d
...
+                return row[column['name'].lower()]
+            elif not self.ignore_missing:
+                raise GOBException(f"Missing column {column['name'].lower()} in query result")
+            else:
+                return None
+
+        def process_row(row):
+            return [process_col(col, row) for col in columns]
+
+        return [self._prepare_row(process_row(row), columns) for row in rows]
```
Comment (RoelWKramer) [inline src/gobbagextract/selector/_selector.py:66]: if _prepare_row() is a member of this class and only used in this method, process_row and process_col should be that as well for consistency. imho.

File: src/tests/conftest.py [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+@pytest.fixture
+def database(app_dir: Path, recreate_database) -> Generator[Session, None, None]:
+    """Fixture which sets up the database, returns a db session.
+
+    :param app_dir: path to current application source.
+    :param recreate_database: fixture which resets the test database.
+    :return: a generator which yields a db session.
+    """
+    test_db_name = recreate_database
+    DATABASE_CONFIG['database'] = test_db_name
```
Comment (RoelWKramer) [inline src/tests/conftest.py:80]: double quotes

File: src/tests/extract_config/test_extract_config.py [hunk 15]
```diff
@@ -1,20 +1,45 @@
...
+from gobbagextract.extract_config import extract_config
 
-datafile = os.path.join(os.path.dirname(__file__), '../../gobbagextract/data/bag.panden.weesp.json')
 
+class TestExtractConfig:
 
-class TestExtractConfig(TestCase):
+    def test_get_extract_definition(self, tests_dir):
+        os.environ["BAG_DATA_CONFIG"] = str(tests_dir / "fixtures" / "bag_data")
+        gemeente = '0457'
```
Comment (RoelWKramer) [inline src/tests/extract_config/test_extract_config.py:15]: double quotes, not mixed.

File: src/tests/test_main.py [hunk 16]
```diff
@@ -83,19 +71,15 @@ def test_handle_bag_extract_message(
         result_msg = {
             'dataset_file': 'data/somefile.json',
             'header': {
-                'application': 'APP NAME',
-                'catalogue': 'CAT',
-                'dataset_file': 'data/fromheader.json',
-                'entity': 'ENT',
-                'source': 'Some name',
+                'application': 'BAGExtract',
```
Comment (RoelWKramer) [inline src/tests/test_main.py:74]: do not mix quote types

File: src/utils/test_kadaster_url.py [hunk 17]
```diff
@@ -0,0 +1,30 @@
+import datetime
```
Comment (RoelWKramer) [inline src/utils/test_kadaster_url.py:1]: is this file still required?

File: src/gobbagextract/selector/_to_postgres.py [hunk 18]
```diff
@@ -1,20 +1,27 @@
...
+        :return: transformed row
         """
         for idx, val in enumerate(row):
             if columns[idx]['type'] in ["JSON", "JSONB"]:
                 row[idx] = Json(val)
         return row
 
-    def _write_rows(self, table: str, values, columns: Optional[List[str]] = None):
-        self._dst_datastore.write_rows(table, values, columns)
+    def _write_rows(self, table: str, values, columns: list[dict[str, str]]) -> int:
```
Comment (RoelWKramer) [inline src/gobbagextract/selector/_to_postgres.py:22]: Prepares and write**s** values to a postgresql database and returns number of rows.

File: src/utils/test_kadaster_url.py (hunk 17 above)
Comment (roel-devries) [reply src/utils/test_kadaster_url.py:1]: just for testing the download service

**Review by roel-devries — APPROVED** [review]
